tikv/pd
# election: fix the etcd transaction failure that may be triggered during PD leader campaign

Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What problem does this PR solve?

Close #3849.

### What is changed and how it works?

Delete the leader key directly rather than checking the leader value first. `LeaderTxn` will make this etcd transaction check `ls.leaderValue` before doing the deleting work, however, the current PD is not a leader yet so its `ls.leaderValue` will always be empty which will cause the etcd transaction to fail and PD will not campaign a new leader until the lease of previous leader key become expired.

https://github.com/tikv/pd/blob/c1f312845ca929242b0967ef76d799f38cb3e261/server/election/leadership.go#L134-L143

### Check List



Tests

- Manual test (add detailed scripts or steps below)

Repeat the #3849. Now the re-campaigning will happen smoothly.

```shell
[2021/07/09 01:18:42.336 +08:00] [WARN] [member.go:174] ["the pd leader has not changed, delete and campaign again"] [old-pd-leader="name:\"pd\" member_id:3474484975246189105 peer_urls:\"http://127.0.0.1:2380\" client_urls:\"http://127.0.0.1:2379\" "]
[2021/07/09 01:18:42.337 +08:00] [INFO] [server.go:1197] ["start to campaign pd leader"] [campaign-pd-leader-name=pd]
[2021/07/09 01:18:42.341 +08:00] [INFO] [lease.go:60] ["lease granted"] [lease-id=2463885191939534603] [lease-timeout=3] [purpose="pd leader election"]
[2021/07/09 01:18:42.346 +08:00] [INFO] [leadership.go:111] ["check campaign resp"] [resp="{\"header\":{\"cluster_id\":15674009460217235928,\"member_id\":3474484975246189105,\"revision\":21,\"raft_term\":3},\"succeeded\":true,\"responses\":[{\"Response\":{\"ResponsePut\":{\"header\":{\"revision\":21}}}}]}"]
[2021/07/09 01:18:42.350 +08:00] [INFO] [leadership.go:120] ["write leaderData to leaderPath ok"] [leaderPath=/pd/6982606201821179194/leader] [purpose="pd leader election"]
[2021/07/09 01:18:42.350 +08:00] [INFO] [server.go:1223] ["campaign pd leader ok"] [campaign-pd-leader-name=pd]
```

### Release note

```release-note
None.
```
